# Bagian A — Python dan Lingkungan Kerja Data Science

**Fundamen Sains Data: Pengantar Python untuk Sains Data, Probabilitas, dan Statistika**
Program Studi Informatika, Program Sarjana, Universitas Islam Indonesia

Notebook ini memuat seluruh kode pada Bagian A modul. Cell yang khusus Google Colab
(`google.colab`, `kagglehub`) dibungkus `try/except` sehingga notebook **tetap berjalan
tanpa galat** di luar Colab: di Colab cell dijalankan seperti pada modul, di Jupyter/VS Code
cell hanya menampilkan keterangan singkat.

---
## A.1 Python dalam data science

Pekerjaan data science umumnya mengikuti alur yang berulang: mengumpulkan data, membersihkan
dan menyiapkannya, mengeksplorasi dengan statistik dan grafik, memodelkan, lalu
mengomunikasikan hasil. Python dipakai di seluruh tahapan tersebut karena sintaksnya mudah
dibaca, komunitasnya besar, dan terutama karena ekosistem library (kumpulan kode siap pakai)
yang luas. Modul ini berfokus pada tahap eksplorasi data dan dasar probabilitas yang menjadi
fondasi pemodelan.

**Tabel 1. Library utama Python untuk data science**

| Library | Fungsi utama | Contoh penggunaan di modul |
|---|---|---|
| NumPy | Perhitungan numerik cepat dengan array | Rata-rata, varians, simulasi acak |
| pandas | Mengolah data tabel (baris dan kolom) | Membuat tabel nilai mahasiswa, memfilter data |
| Matplotlib | Membuat grafik | Histogram, boxplot, kurva distribusi |
| SciPy | Fungsi ilmiah, termasuk distribusi probabilitas | Distribusi binomial dan normal |

Selain keempatnya, **scikit-learn** adalah library untuk machine learning. Pemodelan dengan
scikit-learn tidak dibahas di sini, tetapi dasar yang dipelajari di modul ini menjadi bekal
untuk menggunakannya.

---
## A.2 Notebook: Jupyter, Google Colab, dan VS Code

Data science umumnya dikerjakan di **notebook**, yaitu dokumen interaktif yang menggabungkan
kode, hasil eksekusi, grafik, dan teks penjelasan. Berbeda dengan file script `.py` yang
dijalankan sekaligus dari awal sampai akhir, file notebook `.ipynb` terdiri atas **cell** yang
dapat dijalankan satu per satu, sehingga hasil setiap langkah langsung terlihat. Ada dua jenis
cell:

- **Code cell** berisi kode Python, dan hasilnya muncul tepat di bawah cell.
- **Text cell (Markdown cell)** berisi penjelasan yang ditulis dengan Markdown, format teks
  sederhana seperti `# Judul` atau `**tebal**`.

Kode di notebook dijalankan oleh **runtime** (istilah Colab) atau **kernel** (istilah Jupyter
dan VS Code), yaitu proses Python aktif yang menyimpan semua variabel dan library yang sudah
diimpor selama sesi berjalan.

Tiga lingkungan notebook yang umum dipakai adalah:

- **Jupyter Notebook**, lingkungan notebook interaktif berbasis browser yang dipasang di
  komputer sendiri.
- **Google Colab**, layanan notebook Jupyter yang berjalan di server Google. Tidak
  memerlukan instalasi, library utama sudah tersedia, dan notebook disimpan di Google Drive.
- **VS Code**, editor kode yang mendukung script maupun notebook melalui ekstensi Python dan
  Jupyter.

Modul ini memakai **Google Colab** sebagai lingkungan praktik utama karena hasilnya sama di
semua komputer tanpa perlu instalasi.

---
## A.3 Mengenal antarmuka Google Colab

Antarmuka Google Colab terdiri atas beberapa bagian yang akan dipakai berulang kali di
sepanjang modul (lihat **Gambar 1** pada modul):

1. Nama notebook, yang dapat diklik untuk mengganti nama file.
2. Bilah menu (File, Edit, View, Insert, Runtime, Tools, Help).
3. Tombol **+ Code** dan **+ Text** untuk menambah code cell atau text cell.
4. Ikon panel samping, berisi daftar isi, pencarian, daftar variabel, Secrets, dan berkas (Files).
5. Text cell, berisi judul dan penjelasan yang ditulis dengan Markdown.
6. Code cell beserta tombol jalankan di sisi kirinya.
7. Output, yaitu hasil eksekusi yang tampil tepat di bawah code cell.
8. Indikator runtime. Tombol **Connect** berubah menjadi tanda centang beserta grafik RAM dan
   Disk ketika notebook sudah terhubung ke runtime.
9. Ikon Gemini (berbentuk bintang berkilau) di bilah bawah untuk membuka asisten AI.

Tampilan Colab diperbarui secara berkala, sehingga letak dan nama tombol pada versi terbaru
bisa sedikit berbeda.

---
## A.4 Membuat, menjalankan, dan menyimpan notebook

1. Buka <https://colab.research.google.com> dan masuk dengan akun Google.
2. Pilih **File > New notebook in Drive**, lalu klik nama `Untitled0.ipynb` di kiri atas untuk
   menggantinya, misalnya menjadi `praktik_sains_data.ipynb`.
3. Tulis kode di code cell, lalu jalankan.
4. Tambahkan cell baru dengan **+ Code** atau **+ Text**.
5. Notebook tersimpan otomatis di Google Drive, folder `Colab Notebooks`. Simpan manual dengan
   `Ctrl + S`.
6. Unduh notebook melalui **File > Download > Download .ipynb**.

Sebuah code cell dapat dijalankan dengan beberapa cara:

- Klik tombol jalankan (ikon segitiga di dalam lingkaran) di kiri cell.
- Tekan `Shift + Enter` untuk menjalankan cell lalu pindah ke cell berikutnya, atau `Ctrl + Enter`
  untuk menjalankan tanpa berpindah.
- Pilih **Runtime > Run all** (`Ctrl + F9`) untuk menjalankan seluruh cell dari atas ke bawah.

Contoh code cell sederhana yang sudah dijalankan beserta output-nya (`Gambar 2` pada modul):

In [1]:
# Contoh code cell pertama: menjalankan kode Python sederhana
print("Selamat datang di praktik Sains Data!")
print("Notebook berjalan dengan baik.")

Selamat datang di praktik Sains Data!
Notebook berjalan dengan baik.


---
## A.5 Mengganti jenis runtime

**Runtime** adalah mesin virtual di server Google yang menjalankan kode notebook. Secara
bawaan, runtime memakai **CPU**, dan itu sudah cukup untuk seluruh praktik di modul ini. Untuk
pekerjaan berat seperti melatih model deep learning, runtime dapat diganti agar memakai
akselerator:

1. Pilih menu **Runtime > Change runtime type**.
2. Pada **Runtime type**, biarkan `Python 3`.
3. Pada **Hardware accelerator**, pilih `CPU`, `GPU` (misalnya T4 GPU), atau `TPU`. Sebagian
   pilihan hanya tersedia untuk pengguna Colab berbayar.
4. Klik **Save**.

Mengganti jenis runtime memulai sesi baru, sehingga semua variabel hilang dan seluruh cell perlu
dijalankan ulang. Kuota GPU gratis juga terbatas, jadi pakai akselerator hanya bila memang
diperlukan.

---
## A.6 Mengakses dataset

Analisis data selalu diawali dengan memuat dataset ke notebook. Di Colab, dataset dapat berasal
dari empat sumber:

- File dari komputer sendiri yang diunggah ke penyimpanan sesi.
- File di Google Drive yang dihubungkan ke runtime.
- Dataset publik dari Kaggle melalui fitur **Data explorer**.
- File di internet yang dibaca langsung dari URL.

Setelah file tersedia, isinya dibaca dengan pandas, misalnya `pd.read_csv()` untuk file CSV.

### A.6.1 Mengunggah file dari komputer

File diunggah melalui panel **Files** (ikon folder di panel samping kiri). File yang diunggah
tersimpan di penyimpanan sesi (folder `/content`) dan bersifat **sementara**: file terhapus
ketika runtime berakhir atau dimulai ulang. Unggahan juga dapat dilakukan melalui kode, yang
memunculkan tombol **Choose Files**:

In [2]:
# Hanya berjalan di Google Colab
try:
    from google.colab import files
    import pandas as pd

    unggahan = files.upload()           # pilih file, misalnya nilai.csv
    df_upload = pd.read_csv("/content/nilai.csv")
    print(df_upload.head())
except ImportError:
    print("Cell ini hanya berjalan di Google Colab (modul google.colab tidak tersedia).")
    print("Di Jupyter/VS Code lokal, letakkan nilai.csv di folder kerja lalu:")
    print('    df_upload = pd.read_csv("nilai.csv")')

Cell ini hanya berjalan di Google Colab (modul google.colab tidak tersedia).
Di Jupyter/VS Code lokal, letakkan nilai.csv di folder kerja lalu:
    df_upload = pd.read_csv("nilai.csv")


Untuk menyalin path file yang tepat, klik kanan file di panel **Files** lalu pilih **Copy path**.

### A.6.2 Mengakses file di Google Drive

Berbeda dengan penyimpanan sesi, Google Drive bersifat **permanen**. Drive dihubungkan (mount)
ke runtime dengan ikon **Mount Drive** di panel Files, atau dengan kode berikut. Saat dijalankan,
Colab meminta izin akses ke akun Google:

In [3]:
# Hanya berjalan di Google Colab
try:
    from google.colab import drive
    import pandas as pd

    drive.mount("/content/drive")
    df_drive = pd.read_csv("/content/drive/MyDrive/data/nilai.csv")
    print(df_drive.head())
except ImportError:
    print("Cell ini hanya berjalan di Google Colab (modul google.colab tidak tersedia).")
    print('Di lokal, ganti path menjadi: df_drive = pd.read_csv("data/nilai.csv")')
except Exception as galat:
    print("Mount Google Drive dibatalkan atau gagal:", galat)

Cell ini hanya berjalan di Google Colab (modul google.colab tidak tersedia).
Di lokal, ganti path menjadi: df_drive = pd.read_csv("data/nilai.csv")


Folder `MyDrive` adalah folder utama Google Drive pengguna, sehingga path di atas merujuk ke
file `nilai.csv` di dalam folder `data`. Ganti path tersebut sesuai lokasi file masing-masing.
Cara ini cocok untuk dataset yang dipakai berulang kali, karena tidak perlu diunggah ulang
setiap sesi.

### A.6.3 Memakai dataset publik dari Data explorer

Colab menyediakan **Data explorer** untuk mencari dataset, model, dan kompetisi publik dari
Kaggle (**Gambar 5** pada modul). Dataset Kaggle dimuat dengan library `kagglehub`. Fungsi
`kagglehub.dataset_download()` menerima nama dataset dalam format `pemilik/nama-dataset` dan
mengembalikan lokasi folder tempat file disimpan. Contoh berikut memuat dataset *Netflix Movies
and TV Shows*:

In [4]:
# Hanya berjalan di Google Colab
try:
    import os
    import kagglehub
    import pandas as pd

    lokasi = kagglehub.dataset_download("shivamb/netflix-shows")
    print(os.listdir(lokasi))           # melihat nama file di dalam dataset

    df_netflix = pd.read_csv(os.path.join(lokasi, "netflix_titles.csv"))
    print(df_netflix.head())
except ImportError:
    print("Cell ini hanya berjalan di Google Colab (library kagglehub tidak tersedia).")
    print("Di Colab, library kagglehub sudah terpasang secara bawaan.")

Cell ini hanya berjalan di Google Colab (library kagglehub tidak tersedia).
Di Colab, library kagglehub sudah terpasang secara bawaan.


Colab menyimpan salinan (cache) dataset Kaggle yang populer, sehingga dataset seperti ini
umumnya termuat dalam hitungan detik. Untuk sebagian dataset, Colab dapat meminta login akun
Kaggle terlebih dahulu. Periksa juga lisensi dataset di halaman Kaggle-nya sebelum memakainya
untuk tugas atau publikasi.

### A.6.4 Membaca dataset langsung dari URL

File CSV yang tersedia di internet dapat dibaca langsung oleh pandas tanpa diunduh terlebih
dahulu. Cara ini cocok untuk dataset terbuka, misalnya yang disimpan di GitHub.

**Contoh kasus.** Muat dataset *tips*, data tagihan dan tip pelanggan sebuah restoran, dari
repositori GitHub milik library seaborn.

In [5]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv"
df_tips = pd.read_csv(url)

print(df_tips.shape)
print(df_tips.head())

(244, 7)
   total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4


**Hasil yang diharapkan.**

```text
(244, 7)
   total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4
```

**Interpretasi.** Dataset berisi 244 baris dan 7 kolom. Setiap baris mewakili satu kunjungan,
dengan kolom antara lain total tagihan (`total_bill`), besar tip (`tip`), dan hari kunjungan
(`day`). Untuk file di GitHub, gunakan alamat **raw** (diawali `raw.githubusercontent.com`),
bukan alamat halaman GitHub biasa.

---
## A.7 Urutan eksekusi cell

Hasil notebook bergantung pada **urutan cell dijalankan**, bukan urutan cell ditulis. Variabel
yang dibuat di sebuah cell disimpan di runtime dan baru dapat dipakai cell lain setelah cell
pembuatnya dijalankan.

**Contoh kasus.** Sebuah notebook berisi dua cell. Apa yang terjadi bila cell kedua dijalankan
lebih dulu?

In [6]:
# Cell 1: membuat variabel
nilai = 80

# Cell 2: memakai variabel dari Cell 1
print(nilai + 5)

85


**Hasil yang diharapkan bila Cell 1 dijalankan lebih dulu.**

```text
85
```

**Interpretasi.** Bila Cell 2 dijalankan sebelum Cell 1, atau setelah runtime dimulai ulang,
Python menampilkan `NameError` karena `nilai` belum ada di runtime. Angka di dalam kurung siku
di kiri cell, misalnya `[3]`, menunjukkan urutan eksekusinya. Masalah serupa muncul bila sebuah
cell diubah tetapi tidak dijalankan ulang, karena runtime masih menyimpan nilai lama.

Kesalahan jenis ini sering tidak memunculkan pesan galat sama sekali, sehingga sulit dikenali.
Untuk memastikan notebook benar dari awal, pilih **Runtime > Restart session and run all**.
Perintah ini menghapus semua variabel di runtime, lalu menjalankan seluruh cell dari atas ke
bawah. Lakukan langkah ini sebelum mengumpulkan tugas.

Cell berikut memperagakan `NameError` tersebut secara aman (galat ditangkap, notebook tidak
berhenti):

In [7]:
# Peragaan NameError: variabel yang belum pernah dibuat di runtime
try:
    print(variabel_belum_ada + 5)
except NameError as galat:
    print("NameError:", galat)
    print("Jalankan cell pembuat variabel terlebih dahulu, lalu ulangi cell ini.")

NameError: name 'variabel_belum_ada' is not defined
Jalankan cell pembuat variabel terlebih dahulu, lalu ulangi cell ini.


Runtime Colab juga dapat terputus sendiri bila lama tidak digunakan; kode dan teks tetap
tersimpan, tetapi semua variabel hilang sehingga cell perlu dijalankan ulang.

---
## A.8 Menggunakan Gemini di Colab

Colab memiliki asisten AI bawaan bernama **Gemini**. Gemini memahami isi notebook yang sedang
dibuka, termasuk kode dan variabel yang sudah dijalankan (**Gambar 6** pada modul):

1. Ikon Gemini (bintang berkilau) di bagian bawah notebook. Klik ikon ini untuk membuka atau
   menutup panel Gemini.
2. Tab **Gemini**, yaitu panel percakapan yang terbuka di sisi kanan notebook.
3. Contoh pertanyaan yang dapat diklik langsung, misalnya cara memasang library atau memuat data
   dari Google Drive.
4. Kotak perintah (prompt), tempat menuliskan pertanyaan atau permintaan dalam bahasa
   sehari-hari, termasuk bahasa Indonesia.
5. Pemilih model, untuk memilih versi model Gemini yang dipakai.
6. Tautan **generate** pada code cell kosong, untuk meminta Gemini menuliskan kode langsung di
   cell tersebut.

Beberapa cara memakai Gemini yang berguna untuk praktik di modul ini:

- **Menjelaskan kode.** Contoh perintah: *"Jelaskan apa yang dilakukan cell ini baris demi baris."*
- **Membuat kode.** Contoh perintah: *"Buat histogram kolom nilai dari DataFrame df dengan judul
  dan label sumbu."*
- **Menjelaskan dan memperbaiki galat.** Saat sebuah cell menghasilkan galat, klik tombol
  **Explain error** di bawah pesan galat.
- **Menganalisis data secara bertahap.** *Data Science Agent* dapat menyusun rencana analisis,
  menjalankan kode, dan merangkum temuannya.

Usulan perubahan kode ditampilkan dalam bentuk perbandingan (diff) yang dapat diterima atau
ditolak. Kode dari Gemini tetap perlu dibaca dan diuji sebelum dipakai, karena hasilnya bisa
keliru atau tidak sesuai konteks soal. Untuk latihan dan tugas, gunakan Gemini sebagai **tutor
yang menjelaskan**, bukan sebagai pengganti proses berpikir sendiri, dan ikuti ketentuan dosen
tentang penggunaan AI.

---
## A.9 Alternatif: notebook di VS Code

Notebook yang sama juga dapat dijalankan di VS Code pada komputer sendiri. Panduan lengkap
pemasangan ekstensi, pembuatan notebook, dan pemilihan kernel tersedia di dokumentasi resmi VS
Code: <https://code.visualstudio.com/docs/datascience/jupyter-notebooks>